### PR TITLE
Fix version detection of clang on Apple OSes

### LINKED ISSUE #20518: error: instruction requires: AVX-512 ISA AVX-512 VLX ISA
openssl 3.1.0 fails to build on OS X 10.10.x when using Xcode 7.x; OS X 10.11.x; macOS 10.12.x; macOS 10.13.x; and macOS 10.14.x when using Xcode 10.0 or 10.1. The error is:

```
crypto/bn/rsaz-3k-avx512.s:22:2: error: instruction requires: AVX-512 ISA AVX-512 VLX ISA
 vpxord %ymm0,%ymm0,%ymm0
 ^
```

This was reported to MacPorts here: https://trac.macports.org/ticket/67099

Full build logs of failures on macOS 10.10-10.13, and successes on earlier and later versions, are available here: https://ports.macports.org/port/openssl3/builds/

The problem happens because you are deciding whether the compiler supports avx512 based on faulty criteria:

https://github.com/openssl/openssl/blob/738d43634a5192b1be0869f151682bb8e9157d5a/crypto/modes/asm/aes-gcm-avx512.pl#L62-L64

There, we see you try to use avx512 if the compiler is clang and its version is 7.0 or greater. The problem is that there are two different compilers named clang: the original project's, and Apple's fork which is included in Xcode on macOS. Apple's clang fork uses a different version numbering scheme than the original project's. Support for avx512 may indeed have appeared in the original clang version 7.0 but it evidently did not appear until a later version of Apple's fork of clang.

Based on the table at https://en.wikipedia.org/wiki/Xcode#Toolchain_versions it looks like Apple clang 10.0.1 is the first version that was based on the original project's clang 7.0.0. Apple clang 10.0.0 and earlier were based on original clang 6 and earlier.

Apple clang does identify itself as such in the version string, as in:

```
$ clang -v
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

or:

```
% clang -v
Apple clang version 11.0.3 (clang-1103.0.32.62)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

so it should be fairly easy to modify aes-gcm-avx512.pl to check for the `Apple` prefix and use Apple version numbering in that case. However the version number comparison will need to be enhanced to accommodate the third number. (The existing version number comparison code is wrong anyway since it treats the first two numbers as a floating point number.)

Instead of checking compiler version numbers, it would be more reliable to compile a small program using avx512 and see if it succeeds. If yes, use avx512 support; if not, don't. The way that most configure scripts figure out a system's capabilities.

If you want to continue looking at compiler version numbers and are able to access preprocessor macros, the macro `__apple_build_version__` is set for Apple clang and is not set for original clang. (You can also use the value of `__apple_build_version__` instead of looking at the clang version, if preferred.)

### PR DESCRIPTION
**Pull request by t8m**
Added missing fix from the master branch.

Fixes #20518

### CODE REVIEW

**Comment by t8m** [comment]
Only the second patch is applicable to master branch as well.

File: crypto/bn/asm/rsaz-3k-avx512.pl [hunk 1]
```diff
@@ -48,8 +48,17 @@
     $avx512ifma = ($1==2.11 && $2>=8) + ($1>=2.12);
 }
 
-if (!$avx512 && `$ENV{CC} -v 2>&1` =~ /((?:clang|LLVM) version|.*based on LLVM) ([0-9]+\.[0-9]+)/) {
-    $avx512ifma = ($2>=7.0);
+if (!$avx512 && `$ENV{CC} -v 2>&1`
```
Comment (tom-cosgrove-arm) [inline crypto/bn/asm/rsaz-3k-avx512.pl:51]: Unless this script is included by something else (which I couldn't find) `$avx512` is always going to be falsy here (and also at line 46) - is that intentional?

File: crypto/bn/asm/rsaz-3k-avx512.pl [hunk 2]
```diff
@@ -48,8 +48,17 @@
...
+    =~ /(Apple)?\s*((?:clang|LLVM) version|.*based on LLVM) ([0-9]+)\.([0-9]+)\.([0-9]+)?/) {
```
Comment (tom-cosgrove-arm) [inline crypto/bn/asm/rsaz-3k-avx512.pl:52]: (now that this magic invocation is spreading, it seems like a "compiler-detect" function might be a useful abstraction somewhere)

File: crypto/bn/asm/rsaz-3k-avx512.pl (hunk 1 above)
Comment (ryandesign) [reply crypto/bn/asm/rsaz-3k-avx512.pl:51]: I noticed that too (preexisting problem in the script, not caused by this PR). I assume `$avx512` was intended to be `$avx512ifma`?

File: crypto/bn/asm/rsaz-3k-avx512.pl (hunk 2 above)
Comment (t8m) [reply crypto/bn/asm/rsaz-3k-avx512.pl:52]: > (now that this magic regex is spreading, it seems like a "compiler-detect" function might be a useful abstraction somewhere)

Yeah, but that would be for a different PR

File: crypto/bn/asm/rsaz-3k-avx512.pl (hunk 1 above)
Comment (t8m) [reply crypto/bn/asm/rsaz-3k-avx512.pl:51]: IMO not a big deal, but I can add additional fix to the PR for this - this is on master branch too

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Review by neverpanic — APPROVED** [review]
LGTM

**Review by paulidale — APPROVED** [review]

**Review by levitte — APPROVED** [review]

**Comment by t8m** [comment]
Merge to 3.1 and master (only the two applicable commits) branches. Thank you for the reviews.

**Pull request closed without merging**